q2_fragment_insertion/_format.py

```diff
@@ -6,16 +6,86 @@
 # ----------------------------------------------------------------------------
 
+import re
+
+import ijson
+import skbio
+
 import qiime2.plugin.model as model
+from qiime2.plugin import ValidationError
+
+from q2_types.feature_data import AlignedDNAFASTAFormat
+from q2_types.tree import NewickFormat
 
 
 class PlacementsFormat(model.TextFileFormat):
-    def sniff(self):
-        line = open(str(self)).readline()
+    fields = {'tree', 'placements', 'metadata', 'version', 'fields'}
+
+    def _validate_(self, level):
+        keys_found = set()
 
-        # it's json... but would be very expensive to parse
-        return line[0] == '{'
+        # Can't self.open(mode='rb'), so we defer to the backing pathlib object
+        with self.path.open(mode='rb') as fh:
+            root_element = None
+            for prefix, event, value in ijson.parse(fh):
+                if root_element is None:
+                    if event != 'start_map':
+                        raise ValidationError('Root element of file must be a '
+                                              'JSON object')
+                    else:
+                        root_element = True
+
+                # Skip parsing attributes that could be prohibitively large
+                if prefix.startswith('placements') \
+                        or prefix.startswith('tree'):
+                    continue
+
+                # Restricted to only checking root-level keys
+                if event == 'map_key' and prefix == '':
+                    keys_found.add(value)
+
+        if keys_found != self.fields:
+            raise ValidationError('Expected the following fields: %s, found '
+                                  '%s.' % (sorted(self.fields),
+                                           sorted(keys_found)))
 
 
 PlacementsDirFmt = model.SingleFileDirectoryFormat(
     'PlacementsDirFmt', 'placements.json', PlacementsFormat)
+
+
+class RAxMLinfoFormat(model.TextFileFormat):
+    def _validate_(self, level):
+        sigs = ['This is RAxML version', 'Base frequencies',
+                'Final GAMMA likelihood']
+
+        info = self.path.read_text()
+
+        for sig in sigs:
+            new_sig = sig.replace(r' ', r'\W+')
+            if not re.search(new_sig, info):
+                raise ValidationError('Missing structured content: "%s".'
+                                      % sig)
+
+
+class SeppReferenceFormat(model.DirectoryFormat):
+    alignment = model.File(r'aligned-dna-sequences.fasta',
+                           format=AlignedDNAFASTAFormat)
+    phylogeny = model.File(r'tree.nwk', format=NewickFormat)
+    raxml_info = model.File(r'raxml-info.txt', format=RAxMLinfoFormat)
+
+    def _validate_(self, level):
+        seqs = self.alignment.view(skbio.TabularMSA)
+        tree = self.phylogeny.view(skbio.TreeNode)
+
+        seqs.reassign_index(minter='id')
+        alignment_ids = set(seqs.index)
+        phylogeny_ids = {t.name for t in tree.tips()}
+
```

Comment:
in principle, we should add a way to check if the info file matches the phylogeny - but I don't know how :-/

Reply:
I have some ideas on that, stay tuned.

Reply:
Going to skip this for now, see the new comment, in source.